pkg/fqdn/dnspoller.go

```diff
@@ -0,0 +1,381 @@
+		if !found || len(lookupIPs) == 0 {
+			// We have no data. Do nothing. This will use the most recent data if it
+			// was set on target
+			continue
+		}
+
+		updatedRules := target.updateIPs(lookupIPs)
+
+		// The IPs didn't change. No more to be done for this dnsName
+		if len(updatedRules) == 0 {
+			continue
+		}
+
+		// record which DNS names we used to update rules
+		updatedNames[dnsName] = lookupIPs
+
+		// accumulate rules we need to update with CIDR rules
+		for _, rule := range updatedRules {
+			affectedRules[rule] = append(affectedRules[rule], dnsName)
+		}
+	}
+
+	return affectedRules, updatedNames
+}
+
+// addRule places an api.Rule in the source list.
+// ruleKey must be a unique identifier for the sourceRule
+// newDNSNames and oldDNSNames indicate names that were newly added from this
+// rule, or that were seen in this rule but were already polled for.
+// If newDNSNames and oldDNSNames are both empty, the rule was not added to the
+// poll list.
+func (m *DNSPoller) addRule(ruleKey string, sourceRule *api.Rule) (newDNSNames, oldDNSNames []string) {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, egressRule := range sourceRule.Egress {
+		for _, ToFQDN := range egressRule.ToFQDN {
+			for _, dnsName := range ToFQDN.FQDN {
+				target, dnsNameAlreadyExists := m.loadOrStore(dnsName)
+				target.addRule(ruleKey, sourceRule)
+
+				if dnsNameAlreadyExists {
+					oldDNSNames = append(oldDNSNames, dnsName)
+				} else {
+					newDNSNames = append(newDNSNames, dnsName)
+				}
+			}
+		}
+	}
+
+	return newDNSNames, oldDNSNames
+}
+
+// removeRule removes an api.Rule fomr the source rule set.
+// ruleKey must be a unique identifier for the sourceRule
+// isEmpty indicates no more rules rely on this DNS target
+func (m *DNSPoller) removeRule(ruleKey string, sourceRule *api.Rule) (noLongerPolled []string) {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, egressRule := range sourceRule.Egress {
+		for _, ToFQDN := range egressRule.ToFQDN {
+			for _, dnsName := range ToFQDN.FQDN {
+
+				target, found := m.targets[dnsName]
+				if !found {
+					// no-op if nothing is there
+					return
+				}
+
+				isEmpty := target.removeRule(ruleKey, sourceRule)
+				if isEmpty {
+					noLongerPolled = append(noLongerPolled, dnsName)
+					delete(m.targets, dnsName)
+				}
+			}
+		}
+	}
+
+	return noLongerPolled
+}
+
+// loadOrStore returns the stored object for dnsName, or creates a new one if
+// none exists.
+func (m *DNSPoller) loadOrStore(dnsName string) (target *dnsPollTarget, loaded bool) {
+	target, loaded = m.targets[dnsName]
+	if !loaded {
+		target = &dnsPollTarget{
+			dnsName:     dnsName,
+			IPs:         make(map[string]bool),
+			sourceRules: make(map[string]*api.Rule),
+		}
+		m.targets[dnsName] = target
+	}
+
+	return target, loaded
+}
+
+// dnsPollTarget holds a DNS name that needs to be polled, its most recent
+// data, and the api.Rule that depends on it.
+type dnsPollTarget struct {
+	dnsName     string
+	IPs         map[string]bool
```

Comment:
What's the key and value of this map?

Reply:
Why not use `map[string]struct{}` instead of mapping to a bool to represent a set?